File: SUSTech_Mahjang_Unity_Project/Assets/Scripts/LocalServer/TilePool.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using Assets.Scripts.GameMain;
using Assets.Scripts.Util;

namespace Assets.Scripts.LocalServer
{
	class TilePool
	{
		private const int playerNum = 4;

		private List<Department> departments;

		private TileFactory tileFactory;

		private List<Tile> pool;

		/// <summary>
		/// total number of tiles in pool
		/// </summary>
		public int Count { get => pool.Count; }

		/// <summary>
		/// remained number of tiles in pool
		/// </summary>
		public int remain { get => pool.Count - next; }

        /// <summary>
        /// whether pool is sorted or shuffled
        /// </summary>
        //public bool shuffled { get => this.shuffled; private set => this.shuffled = value; }
        public bool shuffled { get; private set; }

        private int next;

		/// <summary>
		/// Tile pool is the stack where all players draw
		/// tiles from, every game has one tile pool.
		/// </summary>
		/// <param name="departments">
		/// List of department of players' charactors,
		/// there are 4 players in one game, so the
		/// length of <paramref name="departments"/>
		/// should be 4.
		/// </param>
		/// <exception cref="ArgumentException"/>
		public TilePool(List<Department> departments)
		{
			// check parameter
			if (departments.Count != playerNum) throw new ArgumentException(
				string.Format("length should be {0}", playerNum),
				nameof(departments)
			);
			// init
			this.departments = new List<Department>();
			tileFactory = new TileFactory();
			pool = new List<Tile>();
			shuffled = false;
			next = 0;
			// set department
			List<Department> all = new List<Department>(
				(Department[])Enum.GetValues(typeof(Department))
			);
			foreach (Department department in departments)
			{
				if (!this.departments.Contains(department))
				{
					this.departments.Add(department);
					all.Remove(department);
				}
			}
			if (this.departments.Count < 4)
				Algorithm.Shuffle(all);
			while (this.departments.Count < 4)
			{
				departments.Add(all[this.departments.Count]);
			}
			// generate tile pool
			GenerateSpecial();
			GenerateNone();
		}

		/// <summary>
		/// shuffle the tile pool
		/// </summary>
		public void Shuffle()
		{
			Algorithm.Shuffle(pool);
			shuffled = true;
		}

		/// <summary>
		/// draw from a shuffled tile pool
		/// </summary>
		/// <returns>id of dwawn tile, -1 if pool is empty</returns>
		/// <exception cref="AccessViolationException"/>
		public int Draw()
		{
			if (!shuffled)
				throw new AccessViolationException("can't draw from unshuffled tile pool");
			if (pool.Count > next)
			{
				return pool[next++].id;
			}
			else
				return -1;
		}

		/// <summary>
		/// to string for debugging
		/// </summary>
		/// <returns></returns>
		public string PoolToString()
		{
			string s = "";
			s += string.Format("TilePool with {0} Tiles\r\n", pool.Count);
			foreach (Tile tile in pool)
			{
				s += " - " + tile.ToString();
			}
			return s;
		}

		private int CalcReserve(int reserve_0, int reserve_1)
		{
			if (reserve_0 < 0 || reserve_0 > 3)
				throw new ArgumentOutOfRangeException(nameof(reserve_0), reserve_0, "should in range 0~3");
			else if (reserve_1 < 0 || reserve_1 > 15)
				throw new ArgumentOutOfRangeException(nameof(reserve_1), reserve_1, "should in range 0~15");
			return (reserve_0 << (int)Location.Reserve0) + (reserve_1 << (int)Location.Reserve1);
		}

		private void GenerateKing(int reserve_0 = 0, int reserve_1 = 0)
		{
			int reserve = CalcReserve(reserve_0, reserve_1);

			for (int i = 0; i < 4; i++)
				pool.Add(tileFactory.GetTile(
					(int)Special.King << (int)Location.Special +
					i << (int)Location.Unique +
					reserve
				));
		}

		private void GenerateLogo(int reserve_0 = 0, int reserve_1 = 0)
		{
			int reserve = CalcReserve(reserve_0, reserve_1);

			for (int i = 0; i < 4; i++)
				pool.Add(tileFactory.GetTile(
					(int)Special.Logo << (int)Location.Special +
					i << (int)Location.Unique +
					reserve
				));
		}

		private void GenerateSign(int reserve_0 = 0, int reserve_1 = 0)
		{
			int reserve = CalcReserve(reserve_0, reserve_1);

			foreach (int department in departments)
				for (int i = 0; i < 4; i++)
					pool.Add(tileFactory.GetTile(
						(int)Special.Sign << (int)Location.Special +
						i << (int)Location.Unique +
						reserve
					));
		}

		private void GenerateLand(int reserve_0 = 0, int reserve_1 = 0)
		{
			int reserve = CalcReserve(reserve_0, reserve_1);

			foreach (int land in Enum.GetValues(typeof(Land)))
				for (int i = 0; i < 4; i++)
					pool.Add(tileFactory.GetTile(
						(int)Special.Land << (int)Location.Special +
						land << (int)Location.Land +
						i << (int)Location.Unique +
						reserve
					));
		}

		private void GenerateSpecial()
		{
			GenerateKing();
			GenerateLogo();
			GenerateSign();
			GenerateLand();
		}

		private void GenerateNone()
		{
			foreach (Department department in departments)
				for (int i = 1; i < 10; i++)
					for (int j = 0; j < 4; j++)
						pool.Add(tileFactory.GetTile(
							(int)Special.None << (int)Location.Special +
							(int)department << (int)Location.Department +
							i << (int)Location.Sequence +
							j << (int)Location.Unique
						));
		}
	}
}

File: ooadproject/src/main/java/com/example/ooadproject/util/TilePool.java
package com.example.ooadproject.util;

import java.util.*;

public class TilePool {

    /*public static void main(String[] args) {
        List<Tile.Department> departments = new ArrayList<>();
        departments.add(Tile.Department.Math);
        departments.add(Tile.Department.Phy);
        departments.add(Tile.Department.Chem);
        departments.add(Tile.Department.Cse);
        TilePool tilePool = new TilePool(departments);
        System.out.println(tilePool.PoolToString());
    }*/

    private final int playerNum = 4;
    private List<Tile.Department> departments;
    private TileFactory tileFactory;
    private List<Tile> pool;
    private int next;
    private boolean shuffled;

    public int getSize() {
        return pool.size();
    }

    public int getRemain() {
        return pool.size() - next;
    }

    public TilePool(List<Tile.Department> departments) {
        if (departments.size() != playerNum) {
            try {
                throw new Exception("length should be 4");
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        this.departments = new ArrayList<>();
        this.tileFactory = new TileFactory();
        this.pool = new ArrayList<>();
        this.shuffled = false;
        this.next = 0;
        //System.out.println(Tile.Department.values().length);
        List<Tile.Department> all = new ArrayList<>(Arrays.asList(Tile.Department.values()));
        for (Tile.Department d : departments
        ) {

            if (!this.departments.contains(d)) {
                this.departments.add(d);
                all.remove(d);
            }
        }
        //System.out.println(this.departments.size());
        //System.out.println(all.size());
        if (this.departments.size() < 4) {
            Collections.shuffle(all);
        }
        while (this.departments.size() < 4) {
            this.departments.add(all.get(this.departments.size()));
        }
        this.generateNone();
        this.generateSpecial();


    }

    public String PoolToString() {
        String s = "";
        s += "TilePool with " + pool.size() + " Tiles\n";
        for (Tile t : this.pool
        ) {
            s += " - " + t.toString() + "\n";
        }
        return s;
    }

    public void Shuffle() {
        Collections.shuffle(this.pool);
        this.shuffled = true;
    }

    public int Draw(){
        if (!shuffled) {
            this.Shuffle();
        }
        if (this.pool.size() > this.next) {
            return pool.get(next++).getId();
        } else {
            return -1;
        }
    }

    private void generateSpecial() {
        this.generateKing();
        this.generateLogo();
        this.generateSign();
        this.generateLand();
    }

    private void generateKing() {
        int r = this.calcReserve(0, 0);
        for (int i = 0; i < 4; i++) {
            this.pool.add(this.tileFactory.getTile(
                    (Tile.Special.King.value() << Tile.Location.Special.value()) +
                            (i << Tile.Location.Unique.value()) +
                            r
            ));
        }
    }

    private void generateLogo() {
        int r = this.calcReserve(0, 0);
        for (int i = 0; i < 4; i++) {
            this.pool.add(this.tileFactory.getTile(
                    (Tile.Special.Logo.value() << Tile.Location.Special.value()) +
                            (i << Tile.Location.Unique.value()) +
                            r
            ));
        }
    }

    private void generateSign() {
        int r = this.calcReserve(0, 0);
        int i = 0;
        for (Tile.Department d : this.departments) {
            this.pool.add(this.tileFactory.getTile(
                    (Tile.Special.Sign.value() << Tile.Location.Special.value()) +
                            (d.value() << Tile.Location.Department.value()) +
                            r
            ));
        }
    }

    private void generateLand() {
        int r = this.calcReserve(0, 0);
        for (Tile.Land l: Tile.Land.values()
             ) {
            for (int i = 0; i < 4; i++) {
                this.pool.add(this.tileFactory.getTile(
                        (Tile.Special.Land.value() << Tile.Location.Special.value()) +
                                (l.value() << Tile.Location.Land.value())+
                                (i << Tile.Location.Unique.value()) +
                                r
                ));
            }
        }

    }

    private int calcReserve(int r0, int r1) {

        return (r0 << Tile.Location.Reserve0.value()) + (r1 << Tile.Location.Reserve1.value());
    }

    private void generateNone() {
        for (Tile.Department d : this.departments
        ) {
            for (int i = 1; i < 10; i++) {
                for (int j = 0; j < 4; j++) {
                    int temp = (
                            (Tile.Special.None.value() << Tile.Location.Special.value()) +
                                    (d.value() << Tile.Location.Department.value())+
                                    (i << Tile.Location.Sequence.value()) +
                                    (j << Tile.Location.Unique.value())
                    );
                    //System.out.println(temp);
                    this.pool.add(this.tileFactory.getTile(temp));
                }
            }
        }
    }


}
